src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicApi.cs

```diff
@@ -0,0 +1,253 @@
+        internal readonly MsQuicNativeMethods.StreamShutdownDelegate _streamShutdownDelegate;
+        internal readonly MsQuicNativeMethods.StreamSendDelegate _streamSendDelegate;
+        internal readonly MsQuicNativeMethods.StreamReceiveCompleteDelegate _streamReceiveCompleteDelegate;
+        internal readonly MsQuicNativeMethods.StreamReceiveSetEnabledDelegate _streamReceiveSetEnabledDelegate;
+
+        internal readonly MsQuicNativeMethods.SetContextDelegate _setContextDelegate;
+        internal readonly MsQuicNativeMethods.GetContextDelegate _getContextDelegate;
+        internal readonly MsQuicNativeMethods.SetCallbackHandlerDelegate _setCallbackHandlerDelegate;
+
+        internal readonly MsQuicNativeMethods.SetParamDelegate SetParamDelegate;
+        internal readonly MsQuicNativeMethods.GetParamDelegate GetParamDelegate;
+
+        internal void RegistrationOpen(byte[] name)
+        {
+            MsQuicStatusException.ThrowIfFailed(_registrationOpenDelegate(name, out IntPtr ctx));
+            _registrationContext = ctx;
+        }
+
+        internal unsafe uint UnsafeSetParam(
+            IntPtr Handle,
+            uint Level,
+            uint Param,
+            MsQuicNativeMethods.QuicBuffer Buffer)
+        {
+            return SetParamDelegate(
+                Handle,
+                Level,
+                Param,
+                Buffer.Length,
+                Buffer.Buffer);
+        }
+
+        internal unsafe uint UnsafeGetParam(
+            IntPtr Handle,
+            uint Level,
+            uint Param,
+            ref MsQuicNativeMethods.QuicBuffer Buffer)
+        {
+            uint bufferLength = Buffer.Length;
+            byte* buf = Buffer.Buffer;
+            return GetParamDelegate(
+                Handle,
+                Level,
+                Param,
+                &bufferLength,
+                buf);
+        }
+
+        public async ValueTask<MsQuicSecurityConfig> CreateSecurityConfig(X509Certificate certificate)
+        {
+            MsQuicSecurityConfig secConfig = null;
+            var tcs = new TaskCompletionSource<object>();
+            uint secConfigCreateStatus = MsQuicConstants.InternalError;
+
+            uint status = _secConfigCreateDelegate(
+                _registrationContext,
+                (uint)QUIC_SEC_CONFIG_FLAG.CERT_CONTEXT,
+                certificate.Handle,
+                null,
+                IntPtr.Zero,
+                SecCfgCreateCallbackHandler);
+
+            MsQuicStatusException.ThrowIfFailed(status);
+
+            void SecCfgCreateCallbackHandler(
+                IntPtr context,
+                uint status,
+                IntPtr securityConfig)
+            {
+                secConfig = new MsQuicSecurityConfig(this, securityConfig);
+                secConfigCreateStatus = status;
+                tcs.SetResult(null);
+            }
+
+            await tcs.Task.ConfigureAwait(false);
```

Comment:
Just curious, can `ConfigureAwait` thrown an exception (in low memory situations for instance)? If so, you could have a problem cleaning up while the async sec config create is in process.

Reply:
It can't, but I understand the concern of cleaning up the security config. I'll try some things.